source/fuzz/transformation_mutate_pointer.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_mutate_pointer.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationMutatePointer::TransformationMutatePointer(
+    const protobufs::TransformationMutatePointer& message)
+    : message_(message) {}
+
+TransformationMutatePointer::TransformationMutatePointer(
+    uint32_t pointer_id, uint32_t fresh_id,
+    const protobufs::InstructionDescriptor& insert_before) {
+  message_.set_pointer_id(pointer_id);
+  message_.set_fresh_id(fresh_id);
+  *message_.mutable_insert_before() = insert_before;
+}
+
+bool TransformationMutatePointer::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // Check that |fresh_id| is fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  auto* insert_before_inst =
+      FindInstruction(message_.insert_before(), ir_context);
+
+  // Check that |insert_before| is a valid instruction descriptor.
+  if (!insert_before_inst) {
+    return false;
+  }
+
+  // Check that it is possible to insert OpLoad and OpStore before
+  // |insert_before_inst|. We are only using OpLoad here since the result does
+  // not depend on the opcode.
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(SpvOpLoad,
+                                                    insert_before_inst)) {
+    return false;
+  }
+
+  const auto* pointer_inst =
+      ir_context->get_def_use_mgr()->GetDef(message_.pointer_id());
+
+  // Check that |pointer_id| is a result id of a valid pointer instruction.
+  if (!pointer_inst || !IsValidPointerInstruction(ir_context, *pointer_inst)) {
+    return false;
+  }
+
+  // Check that the module contains an irrelevant constant that will be used to
+  // mutate |pointer_inst|.
+  auto constant_id = fuzzerutil::MaybeGetZeroConstant(
```

Comment:
I would add a remark that a later transformation can change this to something more interesting.

Reply:
Fixed in 80d0b78.